### PR TITLE
WFLY-6328 Minimize replication timeouts due to topology changes

### PR DESCRIPTION
**Pull request by pferraro**
WFLY-7410 CommandDispatcher.submitOnCluster() futures can throw NPEs

https://issues.jboss.org/browse/WFLY-6328
https://issues.jboss.org/browse/WFLY-7410

### CODE REVIEW

File: clustering/server/src/main/java/org/wildfly/clustering/server/provider/CacheServiceProviderRegistry.java [hunk 1]
```diff
@@ -139,51 +178,57 @@ void register(Node node, T service) {
...
                     }
                 }
-                if (merged) {
-                    try (Batch batch = this.batcher.createBatch()) {
-                        for (Node node: newNodes) {
-                            // Re-assert services for new members following merge since these may have been lost following split
-                            CacheServiceProviderRegistry.this.getServices(node).forEach(service -> this.register(node, service));
-                        }
+            }
+            if (merged) {
```
Comment (kabir) [inline clustering/server/src/main/java/org/wildfly/clustering/server/provider/CacheServiceProviderRegistry.java]: If the previous try block throws an exception, does this next step have to happen (i.e. should it go in a finally block)(

File: clustering/server/src/main/java/org/wildfly/clustering/server/registry/CacheRegistry.java [hunk 2]
```diff
@@ -97,26 +108,34 @@ public boolean accept(Object key) {
...
-            }
-            this.closeTask.run();
-        });
+        this.cache.removeListener(this);
+        final Node node = this.getGroup().getLocalNode();
+        try (Batch batch = this.batcher.createBatch()) {
+            // If this remove fails, the entry will be auto-removed on topology change by the new primary owner
+            this.cache.getAdvancedCache().withFlags(Flag.IGNORE_RETURN_VALUES, Flag.FAIL_SILENTLY).remove(node);
+        }
+        // Cleanup any unregistered listeners
```
Comment (kabir) [inline clustering/server/src/main/java/org/wildfly/clustering/server/registry/CacheRegistry.java]: Should this be in a finally block?

File: clustering/web/infinispan/src/main/java/org/wildfly/clustering/web/infinispan/session/InfinispanSessionManager.java [hunk 3]
```diff
@@ -156,17 +172,22 @@ public void close() {
...
-        });
+        this.cache.removeListener(this);
+        PrivilegedAction<List<Runnable>> action = () -> this.executor.shutdownNow();
+        WildFlySecurityManager.doUnchecked(action);
+        try {
+            this.executor.awaitTermination(this.cache.getCacheConfiguration().transaction().cacheStopTimeout(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+        this.dispatcher.close();
```
Comment (kabir) [inline clustering/web/infinispan/src/main/java/org/wildfly/clustering/web/infinispan/session/InfinispanSessionManager.java]: finally?

**Review by kabir — COMMENTED** [review]
I've made a few comments about putting stuff after try/catch blocks into a finally block which may or may not be relevant.

File: clustering/server/src/main/java/org/wildfly/clustering/server/registry/CacheRegistry.java (hunk 2 above)
Comment (pferraro) [inline clustering/server/src/main/java/org/wildfly/clustering/server/registry/CacheRegistry.java]: Yeah, it should.

File: clustering/web/infinispan/src/main/java/org/wildfly/clustering/web/infinispan/session/InfinispanSessionManager.java (hunk 3 above)
Comment (pferraro) [inline clustering/web/infinispan/src/main/java/org/wildfly/clustering/web/infinispan/session/InfinispanSessionManager.java]: I don't think ExecutorService.awaitTermination can fail with anything other than an InterruptedException - which is handled above - but a finally block is probably a good idea anyway.
This same code is found in a number of other places - I'll make those use a finally block as well.

File: clustering/server/src/main/java/org/wildfly/clustering/server/provider/CacheServiceProviderRegistry.java (hunk 1 above)
Comment (pferraro) [inline clustering/server/src/main/java/org/wildfly/clustering/server/provider/CacheServiceProviderRegistry.java]: The previous `if (!deadNodes.isEmpty())` logic block is generally mutually exclusive with this `if (merged)` logic block.

**Comment by pferraro** [comment]
@kabir Feedback incorporated.

**Pull request merged**